MiSTer-devel/Linux-Kernel_MiSTer
# Add cpufreq/overclock driver

This PR adds a [cpufreq](https://www.kernel.org/doc/html/v5.15/admin-guide/pm/cpufreq.html) driver for the Cyclone V SOC FPGA as configured in the DE10 Nano/MiSTer, including the _optional_ ability to overclock or underclock the HPS MPU, or optionally even enable CPU frequency scaling on the MPU clock.

The driver has been [tested by some folks on the MiSTer forums](https://misterfpga.org/viewtopic.php?t=4320) for quite some time, and I feel confident enough to create a PR for it.

### Why would we need this?
1. The biggest practical use case is currently to run MUNT at full or near-full speed for MIDI without external hardware, which users have been doing for some time.
2. Any future hybrid emulation efforts that are CPU-bound will be enhanced. Although hybrid emulation is mostly academic now, benchmarks have shown a linear improvement on software emulators as well as some of the initial hybrid Amiga efforts.
3. Games have been shown to load slightly faster if the loading is CPU bound :) 
4. CPU frequency scaling and underclocking itself is of just a little value to MiSTer users; the user can save a bit of power. The use of cpufreq is mostly due to the functionality "belonging" there in the kernel. We can think of the ability to get frequency scaling as a freebie.

### Features
- Fully featured [CPUfreq](https://www.kernel.org/doc/html/v5.15/admin-guide/pm/cpufreq.html) driver for DE10 Nano
- 1.2 GHz and 1 GHz overclock options; 400 MHz underclock option
- Control overclock with scripts from userspace. Maybe we can make it an INI option later; that product decision is separate from the driver though.
- Driver keeps CPU frequency at stock by default

### Technical notes
1. The driver does not touch any of the Cyclone V's registers by default! The kernel's behavior remains the same as before unless user action is taken to change it, making any overclock/underclock functionality 100% optional. The effect of the driver is the MPU is initialized at 800 MHz and stays there unless it's touched by the user - we can verify this by logging. This is because the performance cpufreq governor is selected by default by the kernel config and the maximum frequency is set to 800 MHz on boot, which causes cpufreq to lock the frequency at 800 MHz.
2. The driver uses documented [Clock Manager registers](https://www.intel.com/content/www/us/en/programmable/hps/cyclone-v/hps.html#log1525288024274.html) to read/set the current VCO clock, which affects the MPU clock. It also sets dividers to keep the main, dbgat, and cfgs2fuser0 clocks at default despite a different VCO clock.
3. Safety: Active cooling, or very good passive cooling is highly recommended when overclocking. There is, of course, no temperature sensor on the Cyclone V, although the industrial SKU is very heat tolerant. The [userspace scripts](https://github.com/coolbho3k/MiSTer-Overclock-Scripts) I've made available warn the user about cooling; if a user doesn't use the script and skips the warning, then we assume they're technical enough to know what they're doing :) Anecdotally, from extensive past experience, ARM cores of this generation at this process node do not generally risk significant heat/safety issues when overclocking in this manner.

### Testing
- Build and boot the kernel. 
- Check current CPU frequency: `cat "/sys/devices/system/cpu/cpu0/cpufreq/scaling_cur_freq"`
- Set CPU frequency to 1.0 GHz (125%): `echo "1000000" > "/sys/devices/system/cpu/cpu0/cpufreq/scaling_max_freq"`
- Set CPU frequency to 1.2 GHz (150%): `echo "1200000" > "/sys/devices/system/cpu/cpu0/cpufreq/scaling_max_freq"`
- Set CPU frequency to 0.4 GHz (50%): `echo "400000" > "/sys/devices/system/cpu/cpu0/cpufreq/scaling_max_freq"`
- Set CPU frequency to 0.8 GHz (stock): `echo "800000" > "/sys/devices/system/cpu/cpu0/cpufreq/scaling_max_freq"`

I've put the driver through many hours of transition testing - randomly transitioning between frequencies - with good luck.

### Power Consumption
With a few USB devices and a fan plugged in, on the main MiSTer menu, I measured the total system power consumption:
```
1200 MHz (150% overclock): 1.254 A ~= 6.27 W
800 MHz (stock): 1.223 A ~= 6.12 W
400 MHz (50% underclock): 1.195 A ~= 5.98 W
```
